File: CodeLibrary/TabHost/BottomNavigationView/AndroidJavaDesignBottomNavigationView/app/src/main/java/com/jackshi/androidjavatabhostfordesign/MainActivity.java
package com.jackshi.androidjavatabhostfordesign;

import android.support.annotation.NonNull;
import android.support.design.bottomnavigation.LabelVisibilityMode;
import android.support.design.internal.BottomNavigationItemView;
import android.support.design.internal.BottomNavigationMenuView;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.jackshi.androidjavatabhostfordesign.fragment.FindFragment;
import com.jackshi.androidjavatabhostfordesign.fragment.HomeFragment;
import com.jackshi.androidjavatabhostfordesign.fragment.MeFragment;
import com.jackshi.androidjavatabhostfordesign.fragment.MemberFragment;
import com.jackshi.androidjavatabhostfordesign.fragment.MessageFragment;
import com.jackshi.androidjavatabhostfordesign.utils.FragmentChangeManager;
import com.jackshi.androidjavatabhostfordesign.utils.MessageEvent;
import com.jackshi.androidjavatabhostfordesign.utils.MyPagerAdapter;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {
    FrameLayout mFrameLayout;
    ViewPager mViewPager;
    BottomNavigationView mBottomNavigationView;
    ArrayList<Fragment> mList;
    FragmentChangeManager mChangeManager;
    MenuItem mMenuItem;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        //当前api版本28,android9.0,使用design:28.0.0的BottomNavigationView的注意事项
        /**
         * 缺点:
         *  1.页面数最大是5,不能超过5否则会崩溃
         *  2.无角标设置api支持,需要自己单独加载布局,并且角标的位置不可控需要自己调整位置
         *  3.设置图标大小、调整导航栏高度及图标和文字之间的间距需要重写dimens
         *      dimens:design_bottom_navigation_margin、design_bottom_navigation_icon_size、design_bottom_navigation_height
         *优点:
         *  1.有单独的重复点击事件(setOnNavigationItemReselectedListener),可设置再次点击方法,可以实现再次点击刷新列表功能
         *  2.自带部分图标切换动画(如水波纹效果,缩放等)
         */
        initView();
        initData();
        initClick();
    }

    private void initView() {
        mList = new ArrayList<>();
        mFrameLayout = findViewById(R.id.activity_main_fLayout);
        mViewPager = findViewById(R.id.activity_main_viewPage);
        mBottomNavigationView = findViewById(R.id.activity_main_bottomNavigationView);

        //水平平移动画效果
        mBottomNavigationView.setItemHorizontalTranslationEnabled(true);
        //设置item状态文字显示
        mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_AUTO);
        EventBus.getDefault().register(this);
    }

    private void initData() {
        HomeFragment homeFragment = HomeFragment.newInstance("home", "home");
        FindFragment findFragment = FindFragment.newInstance("find", "find");
        MessageFragment messageFragment = MessageFragment.newInstance("message", "message");
        MeFragment meFragment = MeFragment.newInstance("me", "me");
        MemberFragment memberFragment = MemberFragment.newInstance("member", "member");
        mList.add(homeFragment);
        mList.add(findFragment);
        mList.add(messageFragment);
        mList.add(memberFragment);
        mList.add(meFragment);

        //获取整个的NavigationView
        BottomNavigationMenuView menuView = (BottomNavigationMenuView) mBottomNavigationView.getChildAt(0);
        //这里就是获取所添加的每一个Tab(或者叫menu)，
        View tab = menuView.getChildAt(3);
        BottomNavigationItemView itemView = (BottomNavigationItemView) tab;
        //加载我们的角标View，新创建的一个布局
        View badge = LayoutInflater.from(this).inflate(R.layout.badge_layout, menuView, false);
        //添加到Tab上
        itemView.addView(badge);
        TextView tv = badge.findViewById(R.id.badge_layout_tv_msg_count);
        tv.setText("" + 50);

        //1.FrameLayout+BottomNavigationView实现
        mFrameLayout.setVisibility(View.VISIBLE);
        mViewPager.setVisibility(View.GONE);
        mChangeManager = new FragmentChangeManager(getSupportFragmentManager(), R.id.activity_main_fLayout, mList);

//       //2.ViewPager+BottomNavigationView实现
//        mFrameLayout.setVisibility(View.GONE);
//        mViewPager.setVisibility(View.VISIBLE);
//        mViewPager.setAdapter(new MyPagerAdapter(getSupportFragmentManager(), mList, mMenuItem));
//        mViewPager.setCurrentItem(0);
    }

    private void initClick() {
        mBottomNavigationView.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
                mMenuItem = menuItem;
                switch (menuItem.getItemId()) {
                    case R.id.item_bottom_1:
                        mChangeManager.setFragments(0);
//                        mViewPager.setCurrentItem(0);
                        break;
                    case R.id.item_bottom_2:
                        mChangeManager.setFragments(1);
//                        mViewPager.setCurrentItem(1);
                        break;
                    case R.id.item_bottom_3:
                        mChangeManager.setFragments(2);
//                        mViewPager.setCurrentItem(2);
                        break;
                    case R.id.item_bottom_4:
                        mChangeManager.setFragments(3);
//                        mViewPager.setCurrentItem(3);
                        break;
                    case R.id.item_bottom_5:
                        mChangeManager.setFragments(4);
//                        mViewPager.setCurrentItem(4);
                        break;
                    default:
                        break;
                }
                return true;
            }
        });
        mBottomNavigationView.setOnNavigationItemReselectedListener(new BottomNavigationView.OnNavigationItemReselectedListener() {
            @Override
            public void onNavigationItemReselected(@NonNull MenuItem menuItem) {
                switch (menuItem.getItemId()) {
                    case R.id.item_bottom_1:
                        Toast.makeText(MainActivity.this, "homeRe", Toast.LENGTH_SHORT).show();
                        break;
                    case R.id.item_bottom_2:
                        Toast.makeText(MainActivity.this, "findRe", Toast.LENGTH_SHORT).show();
                        break;
                    case R.id.item_bottom_3:
                        Toast.makeText(MainActivity.this, "msgRe", Toast.LENGTH_SHORT).show();
                        break;
                    case R.id.item_bottom_4:
                        Toast.makeText(MainActivity.this, "memberRe", Toast.LENGTH_SHORT).show();
                        break;
                    case R.id.item_bottom_5:
                        Toast.makeText(MainActivity.this, "meRe", Toast.LENGTH_SHORT).show();
                        break;
                    default:
                        break;
                }
            }
        });

//        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
//            @Override
//            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
//            }
//
//            @Override
//            public void onPageSelected(int position) {
//                if (mMenuItem != null) {
//                    mMenuItem.setChecked(false);
//                } else {
//                    mBottomNavigationView.getMenu().getItem(0).setChecked(false);
//                }
//                mMenuItem = mBottomNavigationView.getMenu().getItem(position);
//                mMenuItem.setChecked(true);
//            }
//
//            @Override
//            public void onPageScrollStateChanged(int state) {
//            }
//        });
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onMessageEvent(MessageEvent messageEvent) {
        switch (messageEvent.getType()) {
            case 0:
                if (messageEvent.getFlag()) {
                    //移除选中item水波纹的效果
                    mBottomNavigationView.setItemBackground(null);
                }
                break;
            case 1:
                //改变选中非选中文字大小来达到移除缩放效果
                if (messageEvent.getFlag()) {
                    mBottomNavigationView.setItemTextAppearanceInactive(R.style.bottom_normal_text);
                    mBottomNavigationView.setItemTextAppearanceActive(R.style.bottom_normal_text);
                } else {
                    mBottomNavigationView.setItemTextAppearanceInactive(R.style.bottom_normal_text);
                    mBottomNavigationView.setItemTextAppearanceActive(R.style.bottom_selected_text);
                }
                break;
            case 2:
                if (messageEvent.getPosition() == 0) {
                    mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_AUTO);
                } else if (messageEvent.getPosition() == 1) {
                    mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_SELECTED);
                } else if (messageEvent.getPosition() == 2) {
                    mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_LABELED);
                } else {
                    mBottomNavigationView.setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_UNLABELED);
                }
                break;
            case 3:
                if (messageEvent.getFlag()) {
                    mBottomNavigationView.setItemHorizontalTranslationEnabled(false);
                } else {
                    mBottomNavigationView.setItemHorizontalTranslationEnabled(true);
                }
                break;
            default:
                break;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }
}

File: CodeLibrary/TabHost/BottomNavigationView/AndroidXkotlinBottomNavigationView/app/src/main/java/com/jackshi/androidxkotlinbottomnavigationview/MainActivity.kt
package com.jackshi.androidxkotlinbottomnavigationview

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.bottomnavigation.LabelVisibilityMode
import com.jackshi.androidxkotlinbottomnavigationview.fragment.*
import com.jackshi.androidxkotlinbottomnavigationview.utils.FragmentChangeManager
import com.jackshi.androidxkotlinbottomnavigationview.utils.MessageEvent
import com.jackshi.androidxkotlinbottomnavigationview.utils.MyPagerAdapter
import kotlinx.android.synthetic.main.activity_main.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

class MainActivity : AppCompatActivity() {
    private var mList: ArrayList<Fragment> = ArrayList()
    private lateinit var mMenuItem: MenuItem
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        //当前api版本30,android11,androidx、kotlin版本,使用google.android.material的BottomNavigationView的注意事项
        /**
         * 缺点:
         *  1.页面数最大是5,不能超过5否则会崩溃
         *  2.设置图标大小、调整导航栏高度及图标和文字之间的间距需要重写dimens
         *      dimens:design_bottom_navigation_margin、design_bottom_navigation_icon_size、design_bottom_navigation_height
         *优点:
         *  1.有单独的重复点击事件(setOnNavigationItemReselectedListener),可设置再次点击方法,可以实现再次点击刷新列表功能
         *  2.自带部分图标切换动画(如水波纹效果,缩放等)
         *  3.角标设置api支持
         */
        initView()
        initData()
        initClick()
    }

    private fun initView() {
        activity_main_bottomNavigationView.labelVisibilityMode =
            LabelVisibilityMode.LABEL_VISIBILITY_AUTO
        activity_main_bottomNavigationView.isItemHorizontalTranslationEnabled = true
        var mBadge1 = activity_main_bottomNavigationView.getOrCreateBadge(R.id.item_bottom_1)
        var mBadge2 = activity_main_bottomNavigationView.getOrCreateBadge(R.id.item_bottom_2)
        mBadge1.number = 1899
        mBadge1.isVisible = true
        mBadge2.isVisible = true
        EventBus.getDefault().register(this)
    }

    private fun initData() {
        val homeFragment = HomeFragment.newInstance("home", "home")
        val findFragment = FindFragment.newInstance("find", "find")
        val messageFragment = MessageFragment.newInstance("message", "message")
        val memberFragment = MemberFragment.newInstance("member", "member")
        val meFragment = MeFragment.newInstance("me", "me")

        mList.add(homeFragment)
        mList.add(findFragment)
        mList.add(messageFragment)
        mList.add(memberFragment)
        mList.add(meFragment)
    }

    private fun initClick() {
        activity_main_viewPage.visibility = View.GONE
        activity_main_fLayout.visibility = View.VISIBLE
        var mFragmentManager =
            FragmentChangeManager(supportFragmentManager, R.id.activity_main_fLayout, mList)

//        mMenuItem = activity_main_bottomNavigationView.menu.getItem(0)
//        activity_main_bottomNavigationView.menu.getItem(0).isChecked = true
//        activity_main_viewPage.visibility = View.VISIBLE
//        activity_main_fLayout.visibility = View.GONE
//        activity_main_viewPage.adapter = MyPagerAdapter(supportFragmentManager, mList, mMenuItem)
//        activity_main_viewPage.currentItem = 0

        val mOnNavigationItemReselectedListener =
            BottomNavigationView.OnNavigationItemReselectedListener {
                mMenuItem = it
                when (it.itemId) {
                    R.id.item_bottom_1 -> {
                        Toast.makeText(this@MainActivity, "homeRe", Toast.LENGTH_SHORT).show()
                    }
                    R.id.item_bottom_2 -> {
                        Toast.makeText(this@MainActivity, "findRe", Toast.LENGTH_SHORT).show()
                    }
                    R.id.item_bottom_3 -> {
                        Toast.makeText(this@MainActivity, "messageRe", Toast.LENGTH_SHORT).show()
                    }
                    R.id.item_bottom_4 -> {
                        Toast.makeText(this@MainActivity, "memberRe", Toast.LENGTH_SHORT).show()
                    }
                    else -> {
                        Toast.makeText(this@MainActivity, "meRe", Toast.LENGTH_SHORT).show()
                    }
                }
            }

        val mOnNavigationItemSelectedListener =
            BottomNavigationView.OnNavigationItemSelectedListener {
                when (it.itemId) {
                    R.id.item_bottom_1 -> {
                        mFragmentManager.setFragments(0)
//                        activity_main_viewPage.currentItem = 0
                        return@OnNavigationItemSelectedListener true
                    }
                    R.id.item_bottom_2 -> {
                        mFragmentManager.setFragments(1)
//                        activity_main_viewPage.currentItem = 1
                        return@OnNavigationItemSelectedListener true
                    }
                    R.id.item_bottom_3 -> {
                        mFragmentManager.setFragments(2)
//                        activity_main_viewPage.currentItem = 2
                        return@OnNavigationItemSelectedListener true
                    }
                    R.id.item_bottom_4 -> {
                        mFragmentManager.setFragments(3)
//                        activity_main_viewPage.currentItem = 3
                        return@OnNavigationItemSelectedListener true
                    }
                    else -> {
                        mFragmentManager.setFragments(4)
//                        activity_main_viewPage.currentItem = 4
                        return@OnNavigationItemSelectedListener true
                    }
                }
                false
            }

        activity_main_bottomNavigationView.setOnNavigationItemSelectedListener(
            mOnNavigationItemSelectedListener
        )
        activity_main_bottomNavigationView.setOnNavigationItemReselectedListener(
            mOnNavigationItemReselectedListener
        )

//        activity_main_viewPage.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
//
//            override fun onPageScrollStateChanged(state: Int) {}
//
//            override fun onPageScrolled(
//                position: Int,
//                positionOffset: Float,
//                positionOffsetPixels: Int
//            ) {
//            }
//
//            override fun onPageSelected(position: Int) {
//                if (mMenuItem != null) {
//                    mMenuItem.isChecked = false
//                } else {
//                    activity_main_bottomNavigationView.menu.getItem(0).isChecked = false
//                }
//                mMenuItem = activity_main_bottomNavigationView.menu.getItem(position)
//                mMenuItem.isChecked = true
//            }
//        })
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onMessageEvent(messageEvent: MessageEvent) {
        when (messageEvent.type) {
            0 -> if (messageEvent.flag) {
                //移除选中item水波纹的效果
                activity_main_bottomNavigationView.itemBackground = null
            }
            1 ->//改变选中非选中文字大小来达到移除缩放效果
                if (messageEvent.flag) {
                    activity_main_bottomNavigationView.itemTextAppearanceInactive =
                        R.style.bottom_normal_text
                    activity_main_bottomNavigationView.itemTextAppearanceActive =
                        R.style.bottom_normal_text
                } else {
                    activity_main_bottomNavigationView.itemTextAppearanceInactive =
                        R.style.bottom_normal_text
                    activity_main_bottomNavigationView.itemTextAppearanceActive =
                        R.style.bottom_selected_text
                }
            2 -> when {
                messageEvent.position === 0 -> {
                    activity_main_bottomNavigationView.labelVisibilityMode =
                        LabelVisibilityMode.LABEL_VISIBILITY_AUTO
                }
                messageEvent.position === 1 -> {
                    activity_main_bottomNavigationView.labelVisibilityMode =
                        LabelVisibilityMode.LABEL_VISIBILITY_SELECTED
                }
                messageEvent.position === 2 -> {
                    activity_main_bottomNavigationView.labelVisibilityMode =
                        LabelVisibilityMode.LABEL_VISIBILITY_LABELED
                }
                else -> {
                    activity_main_bottomNavigationView.labelVisibilityMode =
                        LabelVisibilityMode.LABEL_VISIBILITY_UNLABELED
                }
            }
            3 -> activity_main_bottomNavigationView.isItemHorizontalTranslationEnabled =
                !messageEvent.flag
            else -> {
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
    }
}